lib/spack/spack/util/environment.py

```diff
@@ -51,9 +57,34 @@ def is_system_path(path):
 
 
 def filter_system_paths(paths):
+    """Return only paths that are not system paths."""
     return [p for p in paths if not is_system_path(p)]
 
 
+def system_paths(paths):
+    """Return only paths that are system paths."""
+    return [p for p in paths if is_system_path(p)]
+
+
+def deprioritize_system_paths(paths):
+    """Put system paths at the end of paths, otherwise preserving order."""
```

Comment:
Instead of having two functions which form the partition, I'd prefer you use a single function and do set operations. This avoids the risk of things getting out of sync and producing unexpected results.

Reply:
@scheibelp  I'm confused: what is the risk of getting out of sync, here? There doesn't seem to me to be a chance unless one of the two functions used (which are defined immediately above) no longer does what the name of said function plainly says what it does. If I use one function and do a set-based partition I'm doing more work and I'm still relying on the one function I do use doing exactly what it says on the tin. What am I missing?

Reply:
That said: done anyway.